### PR TITLE
[Model] Add HiDream-O1-Image support

### PR DESCRIPTION
**Pull request by yixiaoer**
## Purpose

Add HiDream-O1-Image support to vLLM-Omni for #3479 and duplicate issue #3733.

  - add the HiDream-O1-Image pipeline and register the model
  - support tensor parallel inference
  - support Cache-DiT acceleration
  - add offline end-to-end example and pipeline tests

## Test Plan

**vLLM Version:** 0.26.0

**vLLM-Omni Commit:** a09000eba729b2fd1bdcf03ac7dc9a5747bfb31b

**Hardware:** 2 × NVIDIA H100 80GB

## Test Result

- image size: 2048 × 2048
- inference steps: 50
- guidance scale: 5.0
- seed: 42
- dtype: BF16
- tensor parallel size: 2
- no CPU offload

### TP=2 without cache

No cache backend option is passed, so the default `cache_backend=none` path is used.

```bash
"$PYTHON" examples/offline_inference/text_to_image/text_to_image.py \
  --model "$MODEL" \
  --prompt "A cat is sitting next to a sign that says 'HiDream-O1 vLLM-Omni'" \
  --height 2048 \
  --width 2048 \
  --num-inference-steps 50 \
  --guidance-scale 5.0 \
  --seed 42 \
  --enforce-eager \
  --tensor-parallel-size 2 \
  --output "$OUTPUT_DIR/no_cache.png"
```
no_cache_1

### TP=2 with Cache-DiT

```bash
"$PYTHON" examples/offline_inference/text_to_image/text_to_image.py \
  --model "$MODEL" \
  --prompt "A cat is sitting next to a sign that says 'HiDream-O1 vLLM-Omni'" \
  --height 2048 \
  --width 2048 \
  --num-inference-steps 50 \
  --guidance-scale 5.0 \
  --seed 42 \
  --enforce-eager \
  --tensor-parallel-size 2 \
  --cache-backend cache_dit \
  --enable-cache-dit-summary \
  --output "$OUTPUT_DIR/cache.png"
```

cache_1

## Test Result

`Generation latency` below is the example's reported `Total generation time`, which excludes Python startup,
model loading, worker initialization, and process shutdown. 

### `FLASH_ATTN` backend (FA3 on H100)

| Cache backend | Runs | Generation latency per run | Median generation latency | Peak VRAM GPU0/GPU1 |
|---|---:|---:|---:|---:|
| none | 3 | 10.0997 / 10.1692 / 10.0235 seconds | 10.0997 seconds | 10181 / 10133 MiB |
| `cache_dit` | 4 | 5.8139 / 5.9119 / 7.8747 / 5.9617 seconds | 5.9368 seconds | 10709 / 10661 MiB |

Cache-DiT reduced median generation latency by 41.2%, corresponding to a 1.70× speedup. The peak
memory cost was approximately 528 MiB per GPU. All four Cache-DiT runs, including the disclosed
7.8747-second outlier, are included in the median.

### Framework `TORCH_SDPA` cross-check

| Cache backend | Generation latency | Peak VRAM GPU0/GPU1 |
|---|---:|---:|
| none | 13.3662 seconds | 10249 / 10201 MiB |
| `cache_dit` | 7.1784 seconds | 10811 / 10763 MiB |

The independent SDPA check showed a 46.3% latency reduction, corresponding to a 1.86× speedup.

### Cache-DiT runtime evidence

Each Cache-DiT run reported:

```text
Transformer Executed Steps: 100
Cache Steps: 33
CFG Cache Steps: 32
```

The 50-step CFG workload executes the conditional and unconditional branches separately, producing 100 transformer forwards. Cache-DiT therefore recorded 65 cache hits across those 100 forwards.

### Accuracy and output validation

The unchanged no-cache accuracy path passed the checked-in reference thresholds:

```text
SSIM:  0.824608 >= 0.8
PSNR: 19.485493 dB >= 19.0
1 passed, 19 warnings
```

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by RuixiangMa — COMMENTED** [review]
@yixiaoer  Thanks for your contributions， the pipeline implementation does not follow the established code conventions in this repo, please refer to existing models

**Comment by yixiaoer** [comment]
@RuixiangMa Thanks for the feedback. I’ll make the updates.

File: examples/offline_inference/hidream_o1_image/end2end.py
Comment (princepride) [inline examples/offline_inference/hidream_o1_image/end2end.py:1]: Currently, we recommend using `recipe` and `examples/offline_inference/text_to_image` to introduce how to use the model. You can refer to the implementation of other models (such as Bagel).

File: tests/diffusion/models/hidream_o1_image/test_pipeline.py
Comment (princepride) [inline tests/diffusion/models/hidream_o1_image/test_pipeline.py:1]: I think maintaining a smoking test to ensure the precious of output is enough.

File: vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py
Comment (princepride) [inline vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py:1]: Typically we need ensure our model at least support TP. So we need try our best to use vLLM API rather than transformers API. We also can directly use or inherit the module vLLM already have, like `Qwen3VL`

**Review by princepride — CHANGES_REQUESTED** [review]
Over all, the commit structure is OK, I left some recommendations and we also need update `/Users/zhipeng.wang/Documents/GitHub/vllm-omni/docs/user_guide/diffusion_features.md`

**Comment by yixiaoer** [comment]
> Over all, the commit structure is OK, I left some recommendations and we also need update `/Users/zhipeng.wang/Documents/GitHub/vllm-omni/docs/user_guide/diffusion_features.md`

@princepride Thanks for the reply! I submitted the solved commit.

File: vllm_omni/diffusion/registry.py [hunk 1]
```diff
@@ -291,6 +291,11 @@
         "pipeline_hidream_image",
         "HiDreamImagePipeline",
     ),
+    "HiDreamO1ImagePipeline": (
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/registry.py:339]: [P1] Add the model-class mapping used by auto-detection. The official HiDream-O1-Image config declares `Qwen3VLForConditionalGeneration`, so the current resolver returns an unregistered architecture instead of `HiDreamO1ImagePipeline`; this registry entry only helps callers that explicitly pass `model_class_name`. The generic text-to-image/server path will fail before loading. Please add the HiDream signature mapping to both resolver/enrichment paths, or make the example pass the pipeline class explicitly.

**Review by hsliuustc0106 — COMMENTED** [review]
Joint review with #5306: this shared auto-resolution issue remains in the current implementation. I’m leaving this as a comment-only review; no approval or change-request verdict.

**Comment by yixiaoer** [comment]
@princepride can you pls merge this pr, thank you!

File: vllm_omni/diffusion/utils/hf_utils.py [hunk 2]
```diff
@@ -71,6 +71,22 @@ def _looks_like_dreamzero(model_name: str) -> bool:
...
+    try:
+        cfg = config if config is not None else get_hf_file_to_dict("config.json", model_name)
+        if not isinstance(cfg, Mapping) or cfg.get("model_type") != "qwen3_vl":
+            return False
+
+        index = get_hf_file_to_dict("model.safetensors.index.json", model_name)
+        if not isinstance(index, Mapping):
+            return False
+        weight_map = index.get("weight_map")
+        return isinstance(weight_map, Mapping) and "model.final_layer2.linear.weight" in weight_map
```
Comment (princepride) [inline vllm_omni/diffusion/utils/hf_utils.py]: Raw index keys have no `model.` prefix (the loader adds it via ComponentSource(prefix="model.")), so this check is always False and auto-resolution falls through to the unregistered Qwen3VL arch. Drop the `model.` prefix.

File: vllm_omni/diffusion/data.py [hunk 3]
```diff
@@ -1247,7 +1252,17 @@ def enrich_config(self) -> None:
                 model_type = cfg.get("model_type")
                 architectures = cfg.get("architectures") or []
 
-                if model_type == "bagel" or "BagelForConditionalGeneration" in architectures:
+                if model_type == "qwen3_vl":
```
Comment (princepride) [inline vllm_omni/diffusion/data.py]: This forces every `qwen3_vl` config into HiDream (or raises). Gate on a positive HiDream signature, not "any qwen3_vl". Same at line 570.

File: vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+            token_types = token_types.expand(batch_size, -1)
+        token_types = token_types.to(inputs_embeds.device)
+        if token_types.shape != (batch_size, total_seq_len):
+            raise ValueError(
+                f"token_types must have shape {(batch_size, total_seq_len)}, got {tuple(token_types.shape)}"
+            )
+
+        min_value = torch.finfo(inputs_embeds.dtype).min
+        masks = []
+        for batch_idx in range(batch_size):
```
Comment (princepride) [inline vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py]: this S×S mask is rebuilt every step but only depends on `token_types`. Build it once per request.

**Review by princepride — CHANGES_REQUESTED** [review]
PTAL

**Comment by yixiaoer** [comment]
@princepride @hsliuustc0106, thank you! All comments are resolved.

File: vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+        key = key.view(batch_size, seq_len, self.qkv_proj.num_kv_heads, self.head_dim).transpose(1, 2)
+        value = value.view(batch_size, seq_len, self.qkv_proj.num_kv_heads, self.head_dim).transpose(1, 2)
+        query = self.q_norm(query)
+        key = self.k_norm(key)
+        query, key = _apply_rotary_pos_emb(query, key, *position_embeddings)
+        num_kv_groups = self.qkv_proj.num_heads // self.qkv_proj.num_kv_heads
+        key = repeat_kv(key.transpose(1, 2), num_kv_groups).transpose(1, 2)
+        value = repeat_kv(value.transpose(1, 2), num_kv_groups).transpose(1, 2)
+
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py]: [P1] Please route this through `vllm_omni.diffusion.attention.layer.Attention` instead of calling raw `scaled_dot_product_attention`. This bypasses the diffusion attention backend/configuration path and leaves the 2048x2048 text+image sequence on an unvalidated dense SDPA path; the pipeline also passes `use_flash_attn=False`. Please use the framework Attention implementation, or document an approved exception and add high-resolution memory/backend validation.

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+                255,
+            )
+        ).astype(np.uint8)
+
+        return Image.fromarray(arr).convert("RGB")
+
+    return post_process_func
+
+
+class HiDreamO1ImagePipeline(nn.Module, DiffusionPipelineProfilerMixin, ProgressBarMixin):
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py:435]: [P2] This new top-level diffusion pipeline has no Cache-DiT or TeaCache integration. For a 50-step 2048x2048 model, please either add one supported cache acceleration path or explicitly defer it with a tracked follow-up/maintainer exception; documenting both features as unsupported does not satisfy the new-model precheck readiness gate.

**Comment by hsliuustc0106** [comment]
[P1] The Test Result is stale: it runs `examples/offline_inference/hidream_o1_image/end2end.py`, but that file is absent from the current head; the checked-in recipe uses `examples/offline_inference/text_to_image/text_to_image.py`. Please update the command, vLLM-Omni commit, and result summary. The new-model evidence also needs concrete H100 latency and peak-VRAM measurements with hardware, resolution, and step count.

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py (hunk 6 above)
Comment (yixiaoer) [reply vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py:435]: Thank you! The Cache-DiT adapter is now attached to`model.model.language_model`. On 2×H100, 2048×2048, 50 steps, TP=2:
  - `FLASH_ATTN`/FA3 no-cache: 10.0997 s median
  - `FLASH_ATTN`/FA3 + Cache-DiT: 5.9368 s median

**Comment by yixiaoer** [comment]
Thank you @hsliuustc0106! Resolved.

**Comment by yixiaoer** [comment]
Hi @princepride , I updated it. Could you please take a look? Thank you!

**Comment by princepride** [comment]
Sure, I will check it later, thanks for your contribute

**Review by princepride — APPROVED** [review]
LGTM

**Comment by princepride** [comment]
@yixiaoer CI failed, PTAL
